lfsapi/auth.go

```diff
@@ -24,13 +24,13 @@ var (
 // authentication from netrc or git's credential helpers if necessary,
 // supporting basic and ntlm authentication.
 func (c *Client) DoWithAuth(remote string, req *http.Request) (*http.Response, error) {
-	return c.doWithAuth(remote, req, nil)
+	return c.doWithAuth(remote, req, nil, false)
 }
 
-func (c *Client) doWithAuth(remote string, req *http.Request, via []*http.Request) (*http.Response, error) {
+func (c *Client) doWithAuth(remote string, req *http.Request, via []*http.Request, ignoreNetRc bool) (*http.Response, error) {
```

Comment:
To make sure that I'm reading this change correctly: do you intend to mark when `doWithAuth` should and should not use the credentials found in `~/.netrc` upon each request?

If this is the case, it sounds like we start with credentials found in `~/.netrc`, and (if those fail), move on to the credential helper, and then the URL credentials, and so on.

That sounds like a reasonable plan to me. I have a few stylistic remarks regarding the new parameter to this function:

1. Do you think we could name the boolean argument positively? For example, instead of `ignoreNetrc` (nit also, casing should be `Netrc` not `NetRc`), could we name it as `useNetrc`? I think that this makes the conditional on line L123 and L155 easier to grok. Instead of `if !ignoreNetrc`, it would read `if useNetrc`, which is IMHO easier to read.

1. Should this argument be added at all? I usually find that adding a boolean argument to a function is a smell indicating a split in behavior change that should instead be applied via a function extraction. Do you think that this is possible with the current code structure? I'm generally wary of changing code in this file, but I think that removing the 6th return value from `getCreds` might be a worthwhile tradeoff for the risk.

More broadly, though, I'm wondering if this is the right way to go about this change. The credential helper that this function (`c.getCreds`) returns has the following interface:

```go
type CredentialHelper interface {
        Fill(Creds) (Creds, error)
        Reject(Creds) error
        Approve(Creds) error
}
```

and has an implementation that tries multiple `CredentialHelper`s in succession until either (1) one succeeds or (2) all of them fail. That goes by the name of `*CredentialHelpers`. I'm wondering if we could replace:

https://github.com/git-lfs/git-lfs/blob/v2.5.1/lfsapi/auth.go#L174

with an implementation of the `CredentialHelper` interface that fills from a user's `~/.netrc`. In this case, `Accept()` would be a no-op, but `Reject()` could indicate that further calls to `Fill()` for a given hostname should fail immediately. A bad `~/.netrc` file could return an error anytime that it's asked to fill a request.

This way, we could return one `CredentialHelper` (backed by a `*CredentialHelpers`) that would transparently attempt the `~/.netrc` first, and then cascade down through other credential helpers without having to re-call this method and juggle the extra argument.

Reply:
I'm with you on the credential helper approach. I think there are two really clear benefits: the code will be cleaner, and URL credentials will be prioritized above .netrc credentials, which will make us more consistent with git. 

My only concern is that this change will modify auth priorities and the order in which we perform auth operations in this file. I discovered first-hand while making this change that even a simple change in ordering can cause some weird, unexpected behaviours (for example, switching the order of the conditionals at lines [123](https://github.com/git-lfs/git-lfs/pull/3179/files#diff-6d925a0e7b322da76d57d6d1f28fc30cR123) and [127](https://github.com/git-lfs/git-lfs/pull/3179/files#diff-6d925a0e7b322da76d57d6d1f28fc30cR127) causes some requests to fall through to the credential helper when they should be using .netrc instead). I think we'd want to make extra sure the change was safe before it goes into a release.